processing/src/main/java/io/druid/query/extraction/NamespacedExtraction.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.inject.name.Named;
+import com.metamx.common.StringUtils;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.nio.ByteBuffer;
+
+/**
+ * Namespaced extraction is a special case of DimExtractionFn where the actual extractor is pulled from a map of known implementations.
+ * In the event that an unknown namespace is passed, a simple reflective function is returned instead.
+ */
+public class NamespacedExtraction extends DimExtractionFn
+{
+  private static final byte CACHE_TYPE_ID = 0x05;
```

Comment:
Noticing the cache id got me thinking and I think that it's generally not safe to cache things that are namespaced joins.  The cache invalidation semantics are entirely based on knowing when segments change.  A namespace can change at a whim and that should invalidate the cache, so I think what we actually need to do is two things

1. Have `getCacheKey()` return `null`.  I don't remember if the various parts of Druid handle `getCacheKey()` returning null (and essentially just bypassing the cache in that case), but if it doesn't, we'll have to make it capable of doing that.
2. Make sure that we rewrite queries to remove the namespace extraction *before* the `CachingClusteredClient`.  That will mean that for things that have the namespace rewritten out of them, we can still take advantage of the cache.